### PR TITLE
Cluster labels section does not show up#22433

### PR DESCRIPTION
**Pull request by jeswanke**
https://github.com/stolostron/backlog/issues/22433

### CODE REVIEW

**Review by KevinFCormier — CHANGES_REQUESTED** [review]
Please make sure the labels for PlacementRules with `spec.clusterSelector.matchLabels` get substituted into the PR description and that it works for multiple labels. What happens if there are say 5+ labels? It looks like you are trying to handle `spec.clusterSelector.matchExpressions` - the backend does support this, but it looks like the UI has never understood PlacementRules in that form.

Discussion on the labels is also still ongoing.

File: frontend/public/locales/en/translation.json [hunk 1]
```diff
@@ -562,10 +562,15 @@
     "creation.app.section.prePost": "Configure automation for prehook and posthook",
     "creation.app.section.settings": "Settings: Specify application behavior",
     "creation.app.section.techPreview": "Technology Preview",
-    "creation.app.select.existing.placement.rule": "Select existing placement rule",
+    "creation.app.select.existing.placement.rule": "Select an existing placement configuration",
+    "creation.app.settings.clusterSelector": "Create a placement configuration to deploy on clusters with specified labels",
+    "creation.app.local.clusters.only": "{{0}} deploys only to local cluster",
+    "creation.app.clusters.matching": "{{0}} deploys to clusters clusters with these labels: {{1}}",
```
Comment (KevinFCormier) [inline frontend/public/locales/en/translation.json:568]: Word "clusters" is duplicated

File: frontend/public/locales/en/translation.json [hunk 2]
```diff
@@ -562,10 +562,15 @@
...
+    "creation.app.clusters.all.online": "{{0}} deploys to any online clusters and local cluster",
+    "creation.app.select.no.existing.placement.rule": "The selected namespace does not contain any existing placement configurations",
     "creation.app.settings.existingRule": "Select an existing placement configuration",
-    "creation.app.settings.localClusters": "Deploy on local cluster",
-    "creation.app.settings.onlineClusters": "Deploy to all online clusters and local cluster",
+    "creation.app.settings.localClusters": "Deploy only to local cluster",
+    "creation.app.settings.onlineClusters": "Deploy to any online clusters and local cluster",
```
Comment (KevinFCormier) [inline frontend/public/locales/en/translation.json:573]: Keep "all" instead of "any"

File: frontend/public/locales/en/translation.json [hunk 3]
```diff
@@ -562,10 +562,15 @@
     "creation.app.section.prePost": "Configure automation for prehook and posthook",
     "creation.app.section.settings": "Settings: Specify application behavior",
     "creation.app.section.techPreview": "Technology Preview",
-    "creation.app.select.existing.placement.rule": "Select existing placement rule",
+    "creation.app.select.existing.placement.rule": "Select an existing placement configuration",
+    "creation.app.settings.clusterSelector": "Create a placement configuration to deploy on clusters with specified labels",
+    "creation.app.local.clusters.only": "{{0}} deploys only to local cluster",
+    "creation.app.clusters.matching": "{{0}} deploys to clusters clusters with these labels: {{1}}",
+    "creation.app.clusters.all.online": "{{0}} deploys to any online clusters and local cluster",
```
Comment (KevinFCormier) [inline frontend/public/locales/en/translation.json:569]: There should be 2 versions of this similar to `creation.app.settings.onlineClustersOnly` and `creation.app.settings.onlineClusters` that switch based on whether hub self management is enabled.

File: frontend/public/locales/en/translation.json [hunk 4]
```diff
@@ -562,10 +562,15 @@
     "creation.app.section.prePost": "Configure automation for prehook and posthook",
     "creation.app.section.settings": "Settings: Specify application behavior",
     "creation.app.section.techPreview": "Technology Preview",
-    "creation.app.select.existing.placement.rule": "Select existing placement rule",
+    "creation.app.select.existing.placement.rule": "Select an existing placement configuration",
+    "creation.app.settings.clusterSelector": "Create a placement configuration to deploy on clusters with specified labels",
```
Comment (KevinFCormier) [inline frontend/public/locales/en/translation.json:566]: Change to "Deploy application resources on clusters with all specified labels"

File: frontend/public/locales/en/translation.json (hunk 3 above)
Comment (KevinFCormier) [reply frontend/public/locales/en/translation.json:569]: And "any" should be "all"

File: frontend/public/locales/en/translation.json [hunk 5]
```diff
@@ -562,10 +562,15 @@
...
+    "creation.app.local.clusters.only": "{{0}} deploys only to local cluster",
+    "creation.app.clusters.matching": "{{0}} deploys to clusters clusters with these labels: {{1}}",
+    "creation.app.clusters.all.online": "{{0}} deploys to any online clusters and local cluster",
+    "creation.app.select.no.existing.placement.rule": "The selected namespace does not contain any existing placement configurations",
     "creation.app.settings.existingRule": "Select an existing placement configuration",
```
Comment (KevinFCormier) [inline frontend/public/locales/en/translation.json:571]: Please look into why we have this string twice!

**Review by KevinFCormier — COMMENTED** [review]
I added comments for the changes needed. In addition there is this one:
```
"creation.app.placement.rule": "Select clusters for application deployment",
```

**Comment by jeswanke** [comment]
@KevinFCormier  -- for this one: https://github.com/stolostron/console/pull/1800#pullrequestreview-1029299051

matchLabels work as long as the key is unique--if a user wanted to say cloud=aws and azure, they would need matchExpressions--iow they can't do:
  matchLabels:
     cloud: 'AWS'
     cloud: 'AZR'

File: frontend/public/locales/en/translation.json (hunk 3 above)
Comment (KevinFCormier) [reply frontend/public/locales/en/translation.json]: First comment not addressed.

**Comment by KevinFCormier** [comment]
> @KevinFCormier -- for this one: [#1800 (review)](https://github.com/stolostron/console/pull/1800#pullrequestreview-1029299051)
> 
> matchLabels work as long as the key is unique--if a user wanted to say cloud=aws and azure, they would need matchExpressions--iow they can't do: matchLabels: cloud: 'AWS' cloud: 'AZR'

For a PlacementRule with multiple matchLabels:
```
spec:
  clusterSelector:
    matchLabels:
      foo: bar
      good: bye
      "yes": "no"
```

I see:
image

`matchLabels` is not an array, and your code is trying to substitute the first element of the array.

As I mentioned, I don't think matchExpressions has ever been supported in the UI. It might be good to support it, but your code ignores the `operator` value and uses `=` for everything.

File: frontend/src/routes/Applications/CreateApplication/Subscription/controlData/ControlDataPlacement.js [hunk 6]
```diff
@@ -50,142 +65,149 @@ export const loadExistingPlacementRules = () => {
...
+                const clusterConditions = _.get(rule, 'spec.clusterConditions')
+                let selector = enableHubSelfManagement?.active
+                    ? i18n('creation.app.local.clusters.only', [ruleName])
+                    : unavailable
+                if (clusterSelector?.matchExpressions?.length > 0) {
+                    if (clusterSelector.matchExpressions[0]?.key !== 'local-cluster') {
+                        const getLabels = () => {
+                            return clusterSelector.matchExpressions
+                                .map(({ key, operator, values }) => {
+                                    return `${key}${operator === 'In' ? '=' : '!='}${values.join(', ')}`
```
Comment (KevinFCormier) [inline frontend/src/routes/Applications/CreateApplication/Subscription/controlData/ControlDataPlacement.js:102]: This does not account for the `Exists` and `DoesNotExist` values for `operator`.

```
properties:
                  matchExpressions:
                    description: matchExpressions is a list of label selector requirements. The requirements are ANDed.
                    items:
                      description: A label selector requirement is a selector that contains values, a key, and an operator that relates the key and values.
                      properties:
                        key:
                          description: key is the label key that the selector applies to.
                          type: string
                        operator:
                          description: operator represents a key's relationship to a set of values. Valid operators are In, NotIn, Exists and DoesNotExist.
                          type: string
                        values:
                          description: values is an array of string values. If the operator is In or NotIn, the values array must be non-empty. If the operator is Exists or DoesNotExist, the values array must be empty. This array is replaced during a strategic merge patch.
```

File: frontend/public/locales/en/translation.json (hunk 3 above)
Comment (KevinFCormier) [reply frontend/public/locales/en/translation.json]: Fixed now.

**Comment by KevinFCormier** [comment]
Thanks for all the fixes. I have a few other minor concerns.

1. When using an existing placement rule that matches by labels, the label controls are filled in when you first enter the editor (as in the first repository in this screenshot) but not if you change to one that does (as in the 2nd repository in this screenshot).
2. The delete button for the repository does not work with the new styling you added.
3. `operator` value still not handled properly for `matchExpressions`.

image

I have put a hold on this PR because I think @fxiang1 or @chenz4027 should also review this since it's in the app lifecycle UI. I'll leave the remainder of the review to them.

**Review by fxiang1 — APPROVED** [review]
LGTM!

**Review by chenz4027 — APPROVED** [review]
lgtm

**Pull request merged**